0017-sampler-interface.md

```diff
@@ -0,0 +1,282 @@
+circuit1 = QuantumCircuit(9)
+...
+<use 2039 Parameters in this circuit>
+...
+circuit1.measure_all()
+
+# Specify 128 different parameter value sets for circuit1, in a 32x4 shape.
+parameter_values1 = np.random.uniform(size=(32, 4, 2039))
+
+# Our second circuit happens to have no parameters, but three output registers 
+# named alpha, beta, and gamma of various sizes
+alpha = ClassicalRegister(127, "alpha")
+beta = ClassicalRegister(10, "beta")
+gamma = ClassicalRegister(15, "gamma")
+circuit2 = QuantumCircuit(QuantumRegister(127), alpha, beta, gamma)
+...
+<insert gates, measure into the registers, apply dynamic circuit features, etc.>
+...
+
+# invoke a sampler
+sampler = Sampler(..., shots=1024)
+job = sampler.run([
+    (circuit1, parameter_values1),
+    circuit2
+])
+
+# wait for the result
+job.result()
+
+# get a bundle of results for every task
+>> PrimitiveResult(
+>>     SamplerTaskResult(
+>>         DataBundle(meas=BitArray(<num_samples=1024, num_bits=127, shape=(32, 4)>)), 
+>>         <metadata>
+>>     ), 
+>>     SamplerTaskResult(
+>>         DataBundle(
+>>             alpha=BitArray(<num_samples=1024, num_bits=127, shape=()>),
+>>             beta=BitArray(<num_samples=1024, num_bits=10, shape=()>),
+>>             gamma=BitArray(<num_samples=1024, num_bits=15, shape=()>)
+>>         ), 
+>>         <metadata>
+>>     ), 
+>>     metadata=<global_metadata>
+>> )
+```
+
+The above example shows various features of the design.
+Notice the output results are placed in a `DataBundle`, a namespace container described [here](https://github.com/Qiskit/RFCs/pull/53), with one entry for each output of the corresponding task's circuits.
+These circuits only have bit-array outcomes, as allowed by Qiskit, and therefore all results have type `BitArray`, described later.
+For now, consider the following usage examples.
+
+```python
+
+# get result from the second task above
+result = job.result()[1]
+
+# get the entire quasi-probability distribution
+result.get_quasiprobabilities()
+>> QuasiProbabilityDistribution({1351235663: 0.0009765625, 325: 0.0009765625, 45262346723465234523453: 0.0009765625, 35435345235234372457453: 0.0009765625, ...})
+
+# marginalize a couple of indexes in the beta output
+result.data.beta.get_counts(marginalize_idxs=[0, 1])
+>> Counts({"00": 189, "10": 11, "01": 801, "11": 123})
+
+# flip all of the bits in the gamma according to some other compatibly-shaped bit-array
+flips = BitArray(my_flips, num_bits=15)
+flipped_gamma = result.data.gamma ^ flips
+```
+
+If a circuit had a non bit-array outcome, it would use a different type in place of `BitArray`.
+We suggest just using `numpy.ndarrays` with the closest appropriate `dtype` for non-bit-array outputs. 
+The shape of such an array would be `(*task_shape, num_samples, *output_register_shape)`.
+For particular types, like `angle`, if it is for some reason decided that no NumPy type is satisfactory,
+we can add more containers analagous to `BitArray`.
```

Comment:
This is an area where we expect both Qiskit and `Sampler` implementers to be evolving in the near term.  In particular, specific `Sampler` implementations are likely to start supporting different OQ3 output types (bit array vs others) at different rates, and Qiskit will expand its capability to represent outputs, which will then take time for sampler implementations to adapt to.

Is it worth considering how we can make the `SamplerV2` (and `EstimatorV2`, I suppose) easier to expand in the future?  One possible example, having an instance expose a "supported features" attribute, similar to the IBM backends of today, which continues values that have precisely defined semantics (defined _only_ by the base interface) for what a user can expect this sampler to support?  As a very very rough idea, a "feature" could be `"IMPLEMENTATION_LEVEL_1"` and this could mean "circuits that contain a single input qubit register and a single output classical register of the same width, and contain only a sequence of gate calls without classical conditions followed by a sequence of measurements".  This kind of built-in versioning could give us a bit more scope to expand the capabilities of the `Sampler` in the future, and to define the required output types for more dynamic circuits later.

(This isn't something that seems obviously correct in my head at all, I'm just thinking about trying to avoid being in another breaking-change `SamplerV3` situation next year.)

Reply:
I think the proposal is you continue using continuous memory to store other types, but that in the future we may want more convenience wrappers to present that data in a usable/ergonomic fashion.

You comment about allowing implementations to declare feature support is interesting, but I think it requires its own independent design process. So, either it becomes something we can add-on to SamplerV2 as an enhancement, or it will have to wait to SamplerV3.

Reply:
My aim is mostly to avoid needing a `SamplerV3` at all, but pragmatically, it may be too late in the design process to avoid that now; we'd risk doing it wrong if we rushed it in too fast, so it might be better to bite the bullet and know we'll have to do it again in a year's time.

The general principle is around defining expandable interfaces with version negotiation.  Terra's `UnitarySynthesis` plugins do this, for example - we say that all attributes whose names start `supports_` are reserved names, and if they're not present, they default to `False`.  Then when we add new capabilities that might want to be abstracted over, we define a new "feature flag" with particular semantics, and plugins can add it slowly over time, without us needing a new interface.

It's not a silver bullet by any means, and the case of a unitary-synthesis compiler plugin is easier than `Sampler`, but it's the general principle I was getting at.

Reply:
That sounds like a useful pattern. But like you say, I think we have run out of time to add that to the Primitive interfaces in this iteration.